runtime/utils/common/src/aws/smithy/kotlin/runtime/util/net/Text.kt

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.util.net
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * Parses a `host[:port]` pair. IPv6 hostnames MUST be enclosed in square brackets (`[]`).
+ */
+@InternalApi
+public fun String.splitHostPort(): Pair<String, Int?> {
+    val lBracketIndex = indexOf('[')
+    val rBracketIndex = indexOf(']')
+    val lastColonIndex = lastIndexOf(":")
+    val hostEndIndex = when {
+        rBracketIndex != -1 -> rBracketIndex + 1
+        lastColonIndex != -1 -> lastColonIndex
+        else -> length
+    }
+
+    if (lBracketIndex == -1 && rBracketIndex != -1 ||
+        lBracketIndex != -1 && rBracketIndex == -1 ||
+        lBracketIndex > rBracketIndex
+    ) {
+        throw IllegalArgumentException("unmatched [ or ]")
+    }
+    if (lBracketIndex > 0) {
+        throw IllegalArgumentException("unexpected characters before [")
+    }
+    if (rBracketIndex != -1 && rBracketIndex != hostEndIndex - 1) {
+        throw IllegalArgumentException("unexpected characters after ]")
+    }
+
+    val host = if (lBracketIndex != -1) {
+        substring(lBracketIndex + 1 until rBracketIndex)
+    } else {
+        substring(0 until hostEndIndex)
+    }
+    if (lBracketIndex != -1 && rBracketIndex != -1 && !host.isIpv6()) {
+        throw IllegalArgumentException("non-ipv6 host was enclosed in []-brackets")
+    }
+
+    return Pair(
+        host,
+        if (hostEndIndex != -1 && hostEndIndex != length) substring(hostEndIndex + 1).toInt() else null
+    )
+}
+
+/**
+ * Validates a hostname per [RFC 1123](https://www.ietf.org/rfc/rfc1123.txt).
+ */
+@InternalApi
+public fun String.isValidHostname(): Boolean {
+    if (length !in 1..63 || !this[0].isLetterOrDigit()) {
+        return false
+    }
+
+    for (i in 1 until length) {
+        val c = this[i]
+        if (!c.isLetterOrDigit() && c != '-') {
+            return false
+        }
+    }
+
+    return true
+}
+
+@InternalApi
+public fun String.isIpv4(): Boolean {
+    val segments = split('.')
+    return segments.size == 4 && segments.all { (it.toIntOrNull() ?: -1) in 0..255 }
+}
+
+/**
+ * Validates a string as an IPv6 address according to [RFC 4291 §2.2](https://www.rfc-editor.org/rfc/rfc4291.html#section-2.2).
+ * Does NOT validate for [address prefixes (§2.3)](https://www.rfc-editor.org/rfc/rfc4291.html#section-2.3).
+ *
+ * The address MAY include a scope ID which is validated per [RFC 6874](https://www.rfc-editor.org/rfc/rfc6874).
+ */
+@InternalApi
+public fun String.isIpv6(): Boolean {
+    val components = split("%25") // delimiter and scope id are %-encoded
```

Comment:
**Question**: Why are delimiter and scope ID percent-encoded? Does some code handle that for us or is it some invariant of IPv6 addresses?

Reply:
Yeah, this is specified in the RFC for scope ID. https://www.rfc-editor.org/rfc/rfc4007

Reply:
Okay, I see where that doc says [`%` is the delimiter between address and zone ID](https://www.rfc-editor.org/rfc/rfc4007#section-5:~:text=%60%25%27%20is%20a%20delimiter%20character%20to%20distinguish%20between%20%3Caddress%3E%20and%0A%20%20%20%20%20%20%3Czone_id%3E.) and that [such a delimiter may need special handling in URIs](https://www.rfc-editor.org/rfc/rfc4007#section-5:~:text=Also%2C%20the%20format%20for%20non%2Dglobal%20addresses%20might%20conflict%20with%20the%20URI%0A%20%20%20syntax%20%5B13%5D%2C%20since%20the%20syntax%20defines%20the%20delimiter%20character%20%28%60%25%27%29%0A%20%20%20as%20the%20escape%20character.%20%20This%20conflict%20would%20require%2C%20for%20example%2C%0A%20%20%20that%20the%20%3Czone_id%3E%20part%20for%20zone%201%20with%20the%20delimiter%20be%20represented%0A%20%20%20as%20%27%25251%27.). What I don't yet understand is what mechanism will have already percent-encoded the string?

Reply:
Ok you had me tripped up until I realized there are TWO RFCs for this behavior. I linked 4007. We're parsing zone IDs that appear in URIs which means we want to pay attention to https://datatracker.ietf.org/doc/html/rfc6874. The latter is what my code references, I inadvertently discovered and linked it when I went to respond to this question.